src/usr/share/netcheck/networkmanagerhelper.py

```
@@ -0,0 +1,647 @@
+                            str(exception))
+                        raise
+                else:
+                    raise
+
+            except NetworkManagerHelper.ObjectVanished as exception:
+                vanished_symbol_count += 1
+                if vanished_symbol_count >= VANISHED_SYMBOL_MAX_ATTEMPTS:
+                    self.logger.error(
+                        'Object vanished after %d retry attempts. %s: %s',
+                        vanished_symbol_count, type(exception).__name__, str(exception))
+                    raise
+
+        return return_value
+
+    return passthrough_on_reentry
+
+class NetworkManagerHelper(object):
+    """NetworkManagerHelper abstracts away some of the messy details of the NetworkManager
+    D-Bus API. All methods will retry for a bit when encounting exceptions that the
+    NetworkManager API frequently throws.
+    """
+    # TODO: Eventually remove this non-sense once we implement propper logging. (Yes, this is
+    #   all to support our logger as an instance variable.) (gpgmailer issue 18)
+    NetworkManager = None
+    ObjectVanished = None
+
+    def __init__(self, config):
+        """Constructor.
+
+        config: The configuration dictionary constructed during program initialization.
+        """
+
+        self.logger = logging.getLogger(__name__)
+        self.connection_activation_timeout = config['connection_activation_timeout']
+        self.connection_ids = config['connection_ids']
+
+        self.random = random.SystemRandom()
+
+        self._import_network_manager()
+
+    @reiterative
+    def get_all_connection_ids(self):
+        """Returns all connection IDs known to NetworkManager."""
+        connection_ids = []
+        for connection in self.NetworkManager.Settings.ListConnections():
+            connection_ids.append(connection.GetSettings()['connection']['id'])
+
+        return connection_ids
+
+    @reiterative
+    def update_available_connections(self):
```

This is pretty unlikely, but if there are no devices this will just silently do nothing. Is that alright?

---

Yeah, that is fine. A device might get plugged in later.